backend/segmentation/models/microcapsule.py

```
@@ -0,0 +1,245 @@
+            kept.append(i)
+    return kept
+
+
+def _bpe_path() -> str:
+    """Path to the CLIP BPE vocab; download + cache under HF_HOME on first use."""
+    cache_dir = os.environ.get("HF_HOME") or "/tmp"
+    path = os.path.join(cache_dir, "sam3_bpe_simple_vocab_16e6.txt.gz")
+    if not os.path.exists(path):
+        os.makedirs(cache_dir, exist_ok=True)
+        logger.info("Downloading CLIP BPE vocab for SAM 3 ...")
+        urllib.request.urlretrieve(_BPE_URL, path)
```

Ruff (0.15.18)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)